tests/handlers/test_sync.py

```diff
@@ -186,6 +187,139 @@ def test_unknown_room_version(self):
+                create_requester(owner), generate_sync_config(owner)
+            )
+        )
+        self.assertEqual(len(alice_sync_result.joined), 1)
+        self.assertEqual(alice_sync_result.joined[0].room_id, room_id)
+        last_room_creation_event_id = (
+            alice_sync_result.joined[0].timeline.events[-1].event_id
+        )
+
+        # (Pretend that) a remote user Bob joins.
+        moderator = "@bob:example.com"
+        self.get_success(
+            inject_event(
+                self.hs,
+                sender=moderator,
+                prev_event_ids=[last_room_creation_event_id],
+                type=EventTypes.Member,
+                state_key=moderator,
+                room_id=room_id,
+                content={"membership": "join"},
+            )
+        )
+
+        # Alice makes Bob a moderator.
+        room_power_levels = self.helper.get_state(
+            room_id,
+            EventTypes.PowerLevels,
+            tok=owner_tok,
+        )
+        room_power_levels["users"][moderator] = 50
+        moderator_pl_event_id = self.helper.send_state(
+            room_id,
+            EventTypes.PowerLevels,
+            room_power_levels,
+            tok=owner_tok,
+        )["event_id"]
+
+        # We now want to concoct a situation where:
+        # - another local user Eve attempts to join the room
+        # - the homeserver generates a join event with prev_events that precede the ban
+        #   (so that it passes the "are you banned" test)
+        # - but the join event has a stream_ordering after that of the ban (so that it
+        #   comes down /sync after the ban).
+        #
+        # We do this by
+        # 1. Ban eve from the room.
+        # 2. Have eve join the room.
+        #    - mock the prev_events for the join event to precede the ban
+
+        # First, the ban.
+        eve = self.register_user("eve", "password")
+        eve_token = self.login(eve, "password")
+        self.get_success(
+            inject_event(
+                self.hs,
+                sender=moderator,
+                prev_event_ids=[moderator_pl_event_id],
+                type=EventTypes.Member,
+                state_key=eve,
+                room_id=room_id,
+                content={"membership": "ban"},
+            )
+        )
+
+        # We need a sync_token for eve after the ban.
+        eve_requester = create_requester(eve)
+        eve_sync_config = generate_sync_config(eve)
+        eve_sync_after_ban: SyncResult = self.get_success(
+            self.sync_handler.wait_for_sync_for_user(eve_requester, eve_sync_config)
+        )
+        since_token = eve_sync_after_ban.next_batch
+        # Sanity check this sync result. We should be banned from the room.
+        self.assertEqual(len(eve_sync_after_ban.joined), 0)
+        self.assertEqual(len(eve_sync_after_ban.archived), 1)
+        self.assertEqual(eve_sync_after_ban.archived[0].room_id, room_id)
```

Comment:
Why would this even be in the user's sync response if they never joined the room? Doesn't this essentially leak bans to a user?

Reply:
I didn't expect this to show up either.

My recollection of "archived" from when I looked at the sync machinery before Christmas is that nobody really knows what it's meant to do our how it currently works.

> Doesn't this essentially leak bans to a user?

Maybe---I'd want to try this out in a simpler case without any fakery that injects invents though.

Reply:
Seems to be working as expected.